app/report_requests/process_notifications_report.py

```diff
@@ -0,0 +1,161 @@
+import csv
+import sys
+from io import StringIO
+
+from flask import current_app
+from notifications_utils.s3 import (
+    S3_MULTIPART_UPLOAD_MIN_PART_SIZE,
+    s3_multipart_upload_abort,
+    s3_multipart_upload_complete,
+    s3_multipart_upload_create,
+    s3_multipart_upload_part,
+)
+
+from app.constants import ALL_NOTIFICATION_TYPES, ALL_TYPE, NOTIFICATION_REPORT_REQUEST_MAPPING
+from app.dao.notifications_dao import get_notifications_for_service
+from app.dao.report_requests_dao import dao_get_report_request_by_id
+from app.dao.service_data_retention_dao import fetch_service_data_retention_by_notification_type
+
+
+def process_report_request(service_id, report_request_id):
+    report_request = dao_get_report_request_by_id(service_id, report_request_id)
+    notification_type = report_request.parameter["notification_type"]
+    notification_status = report_request.parameter["notification_status"]
+    substituted_types = ALL_NOTIFICATION_TYPES if notification_type == ALL_TYPE else [notification_type]
+
+    page_size = current_app.config.get("REPORT_REQUEST_NOTIFICATIONS_CSV_BATCH_SIZE")
+    s3_bucket = current_app.config["S3_BUCKET_REPORT_REQUESTS_DOWNLOAD"]
+    filename = f"notifications_report/{report_request_id}.csv"
+    part_number = 1
+
+    init_upload_response = s3_multipart_upload_create(s3_bucket, filename)
+    upload_id = init_upload_response["UploadId"]
+    parts = []
+
+    csv_buffer = StringIO()
+    csv_writer = csv.writer(csv_buffer)
+    accumulated_size = 0
+
+    headers = [
+        "Recipient",
+        "Reference",
+        "Template",
+        "Type",
+        "Sent by",
+        "Sent by email",
+        "Job",
+        "Status",
+        "Time",
+        "API key name",
+    ]
+    csv_writer.writerow(headers)
+
+    try:
+        for substituted_type in substituted_types:
+            service_retention = fetch_service_data_retention_by_notification_type(service_id, substituted_type)
```

Comment:
`fetch_service_data_retention_by_notification_type` returns `None` if the service doesn't have custom data retention for the specified notification type. So for that case, the default value of 7 days should be used instead of `service_retention.days_of_retention`.